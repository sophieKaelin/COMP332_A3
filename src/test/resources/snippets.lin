//
// This file is part of COMP332 Assignment 3 2019.
//
// Lintilla, a simple functional programming language.
//
// Â© 2019, Dominic Verity and Anthony Sloane, Macquarie University.
//         All rights reserved.
//
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.
//
// Miscellaneous code snippets referred to in the README and assignment 3 spec.

let y = 1;
print y;
{
    print y;
    let y = 2;
    print y
};
print y;

fn mod(n: int, m: int) -> int {
    n - (n / m) * m
};
    
print(mod(23,10));  // prints `3`, the remainder when `23` is divided by `10`.

let x = 20;

fn print_n_plus_x(n : int) {
    print x + n
};

print_n_plus_x(23);    // prints the value `43`

fn f(a : int) -> int {
    a + 1
};
    
fn g(b : fn(int) -> int) -> int {
    b(5)
};
    
print g(f);

// {
//     let b = {
//         let p = 1;
//         p + 1
//      };
//      print p + 2
// };

print g;

// fn k(n : int, n : int) { let n = 10; let k = 7 };

fn l() {
    let l = 10;
    print l
};

l();

let z = { print 10; 10 } + { print 5; 5 };
print z

