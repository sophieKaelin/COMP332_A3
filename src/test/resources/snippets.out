Program tree:
-------------

Program(
    Vector(
        LetDecl(IdnDef("y"), IntExp(1)),
        PrintExp(IdnExp(IdnUse("y"))),
        Block(
            Vector(
                PrintExp(IdnExp(IdnUse("y"))),
                LetDecl(IdnDef("y"), IntExp(2)),
                PrintExp(IdnExp(IdnUse("y"))))),
        PrintExp(IdnExp(IdnUse("y"))),
        FnDecl(
            IdnDef("mod"),
            Vector(
                ParamDecl(IdnDef("n"), IntType()),
                ParamDecl(IdnDef("m"), IntType())),
            Some(IntType()),
            Block(
                Vector(
                    MinusExp(
                        IdnExp(IdnUse("n")),
                        StarExp(
                            SlashExp(
                                IdnExp(IdnUse("n")),
                                IdnExp(IdnUse("m"))),
                            IdnExp(IdnUse("m"))))))),
        PrintExp(
            AppExp(IdnExp(IdnUse("mod")), Vector(IntExp(23), IntExp(10)))),
        LetDecl(IdnDef("x"), IntExp(20)),
        FnDecl(
            IdnDef("print_n_plus_x"),
            Vector(ParamDecl(IdnDef("n"), IntType())),
            None,
            Block(
                Vector(
                    PrintExp(
                        PlusExp(IdnExp(IdnUse("x")), IdnExp(IdnUse("n"))))))),
        AppExp(IdnExp(IdnUse("print_n_plus_x")), Vector(IntExp(23))),
        FnDecl(
            IdnDef("f"),
            Vector(ParamDecl(IdnDef("a"), IntType())),
            Some(IntType()),
            Block(Vector(PlusExp(IdnExp(IdnUse("a")), IntExp(1))))),
        FnDecl(
            IdnDef("g"),
            Vector(
                ParamDecl(IdnDef("b"), FnType(Vector(IntType()), IntType()))),
            Some(IntType()),
            Block(Vector(AppExp(IdnExp(IdnUse("b")), Vector(IntExp(5)))))),
        PrintExp(AppExp(IdnExp(IdnUse("g")), Vector(IdnExp(IdnUse("f"))))),
        PrintExp(IdnExp(IdnUse("g"))),
        FnDecl(
            IdnDef("l"),
            Vector(),
            None,
            Block(
                Vector(
                    LetDecl(IdnDef("l"), IntExp(10)),
                    PrintExp(IdnExp(IdnUse("l")))))),
        AppExp(IdnExp(IdnUse("l")), Vector()),
        LetDecl(
            IdnDef("z"),
            PlusExp(
                Block(Vector(PrintExp(IntExp(10)), IntExp(10))),
                Block(Vector(PrintExp(IntExp(5)), IntExp(5))))),
        PrintExp(IdnExp(IdnUse("z")))))

Environment out:
----------------

scope
    "x" -> Variable(IntExp(20))
    "mod" -> Function(fn(int,int)->int,Block(Vector(MinusExp(IdnExp(IdnUse(n)),StarExp(SlashExp(IdnExp(IdnUse(n)),IdnExp(IdnUse(m))),IdnExp(IdnUse(m)))))))
    "y" -> Variable(IntExp(1))
    "f" -> Function(fn(int)->int,Block(Vector(PlusExp(IdnExp(IdnUse(a)),IntExp(1)))))
    "print_n_plus_x" -> Function(fn(int)->unit,Block(Vector(PrintExp(PlusExp(IdnExp(IdnUse(x)),IdnExp(IdnUse(n)))))))
    "g" -> Function(fn(fn(int)->int)->int,Block(Vector(AppExp(IdnExp(IdnUse(b)),Vector(IntExp(5))))))
    "l" -> Function(fn()->unit,Block(Vector(LetDecl(IdnDef(l),IntExp(10)), PrintExp(IdnExp(IdnUse(l))))))
    "z" -> Variable(PlusExp(Block(Vector(PrintExp(IntExp(10)), IntExp(10))),Block(Vector(PrintExp(IntExp(5)), IntExp(5)))))

Target tree:
------------

List(
    IInt(1),
    IClosure(
        None,
        List("y"),
        List(
            IVar("y"),
            IPrint(),
            IVar("y"),
            IPrint(),
            IInt(2),
            IClosure(None, List("y"), List(IVar("y"), IPrint())),
            ICall(),
            IVar("y"),
            IPrint(),
            IClosure(
                Some("mod"),
                List("n", "m"),
                List(
                    IVar("n"),
                    IVar("n"),
                    IVar("m"),
                    IDiv(),
                    IVar("m"),
                    IMul(),
                    ISub())),
            IClosure(
                None,
                List("mod"),
                List(
                    IInt(23),
                    IInt(10),
                    IVar("mod"),
                    ICall(),
                    IPrint(),
                    IInt(20),
                    IClosure(
                        None,
                        List("x"),
                        List(
                            IClosure(
                                Some("print_n_plus_x"),
                                List("n"),
                                List(IVar("x"), IVar("n"), IAdd(), IPrint())),
                            IClosure(
                                None,
                                List("print_n_plus_x"),
                                List(
                                    IInt(23),
                                    IVar("print_n_plus_x"),
                                    ICall(),
                                    IClosure(
                                        Some("f"),
                                        List("a"),
                                        List(IVar("a"), IInt(1), IAdd())),
                                    IClosure(
                                        None,
                                        List("f"),
                                        List(
                                            IClosure(
                                                Some("g"),
                                                List("b"),
                                                List(
                                                    IInt(5),
                                                    IVar("b"),
                                                    ICall())),
                                            IClosure(
                                                None,
                                                List("g"),
                                                List(
                                                    IVar("f"),
                                                    IVar("g"),
                                                    ICall(),
                                                    IPrint(),
                                                    IVar("g"),
                                                    IPrint(),
                                                    IClosure(
                                                        Some("l"),
                                                        Nil,
                                                        List(
                                                            IInt(10),
                                                            IClosure(
                                                                None,
                                                                List("l"),
                                                                List(
                                                                    IVar(
                                                                        "l"),
                                                                    IPrint())),
                                                            ICall())),
                                                    IClosure(
                                                        None,
                                                        List("l"),
                                                        List(
                                                            IVar("l"),
                                                            ICall(),
                                                            IInt(10),
                                                            IPrint(),
                                                            IInt(10),
                                                            IInt(5),
                                                            IPrint(),
                                                            IInt(5),
                                                            IAdd(),
                                                            IClosure(
                                                                None,
                                                                List("z"),
                                                                List(
                                                                    IVar(
                                                                        "z"),
                                                                    IPrint())),
                                                            ICall())),
                                                    ICall())),
                                            ICall())),
                                    ICall())),
                            ICall())),
                    ICall())),
            ICall())),
    ICall())

Output:
-------

1
1
2
1
3
43
6
function of arguments (b)
10
10
5
15
